core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java

```diff
@@ -369,6 +369,12 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
             channelReadComplete(ctx);
             return;
         }
+        if (evt instanceof InitiateConnectionShutdown) {
+            // HTTP/1 doesn't support draining that signals clients about connection shutdown but still
+            // accepts in flight requests. Simply destroy KeepAliveHandler which causes next response
+            // to have a "Connection: close" header and connection to be closed after the next response.
+            destroyKeepAliveHandler();
```

Comment:
```suggestion
            destroyKeepAliveHandler();
            return;
```

Reply:
There is one important thing I missed. 
We cannot destroy `NoopKeepAliveHandler` which is set when keep-alive options are disabled.
https://github.com/line/armeria/blob/cc495bcae11a37dfc8a2d2e166dbf8c1f57fb552/core/src/main/java/com/linecorp/armeria/server/HttpServerPipelineConfigurator.java#L455-L466

I think we can add `CloseableKeepAliveHandler` for HTTP/1 that can only have `needToCloseConnection` flag.
Could you also add a test that triggers `InitiateConnectionShutdown` when a keep-alive option is disabled.

Reply:
How about one of those options:

1. Instead of using only `KeepAliveHandler` to decide if "Connection: close" should be written for HTTP/1 - store a boolean inside `ServerHttp1ObjectEncoder`, add a method to set it from [Http1RequestDecoder](https://github.com/line/armeria/blob/c6150f89c93faa08d08c84317ea2d81400a26c2c/core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java#L376) where we handle `InitiateConnectionShutdown` event, then inside the `ServerHttp1ObjectEncoder` check if `keepAliveHandler.needToCloseConnection()` or that new bool is true to write the header. Not sure if there are any downsides, and we avoid adding yet another `KeepAliveHandler` class.

or 

2. Make `NoopKeepAliveHandler` non-singleton, store a boolean inside the `NoopKeepAliveHandler` that'll be used as a return value of `.needToCloseConnection`, set it to `true` when `.destroy` is called? Obviously may increase memory usage / GC pressure.

Reply:
> store a boolean inside ServerHttp1ObjectEncoder

I love this idea. ❤️

Reply:
Done